packages/core/src/typegenTypes.ts

```diff
@@ -0,0 +1,260 @@
+  TTypesMeta extends TypegenConstraint,
+  TEvent extends EventObject,
+  TAction extends BaseActionObject
+> = TTypesMeta extends TypegenEnabled
+  ? TTypesMeta & {
+      indexedActions: IndexByType<TAction>;
+      indexedEvents: IndexByType<TEvent>;
+    }
+  : TypegenDisabled;
+
+export type TypegenMachineOptionsActions<
+  TContext,
+  TResolvedTypesMeta,
+  // @ts-ignore
+  TEventsCausingActions = TResolvedTypesMeta['eventsCausingActions'],
+  // @ts-ignore
+  TIndexedEvents = TResolvedTypesMeta['indexedEvents'],
+  // @ts-ignore
+  TIndexedActions = TResolvedTypesMeta['indexedActions']
+> = {
+  [K in keyof TEventsCausingActions]?:
+    | ActionObject<
+        TContext,
+        // @ts-ignore
+        TIndexedEvents[TEventsCausingActions[K]]
+      >
+    | ActionFunction<
+        TContext,
+        // @ts-ignore
+        TIndexedEvents[TEventsCausingActions[K]],
+        // @ts-ignore
+        TIndexedActions[K]
+      >;
+};
+
+export type TypegenMachineOptionsDelays<
+  TContext,
+  TResolvedTypesMeta,
+  // @ts-ignore
+  TEventsCausingDelays = TResolvedTypesMeta['eventsCausingDelays'],
+  // @ts-ignore
+  TIndexedEvents = TResolvedTypesMeta['indexedEvents']
+> = {
+  [K in keyof TEventsCausingDelays]?: DelayConfig<
+    TContext,
+    // @ts-ignore
+    TIndexedEvents[TEventsCausingDelays[K]]
+  >;
+};
+
+export type TypegenMachineOptionsGuards<
+  TContext,
+  TResolvedTypesMeta,
+  // @ts-ignore
+  TEventsCausingGuards = TResolvedTypesMeta['eventsCausingGuards'],
+  // @ts-ignore
+  TIndexedEvents = TResolvedTypesMeta['indexedEvents']
+> = {
+  [K in keyof TEventsCausingGuards]?: ConditionPredicate<
+    TContext,
+    // @ts-ignore
+    TIndexedEvents[TEventsCausingGuards[K]]
+  >;
+};
+
+export type TypegenMachineOptionsServices<
+  TContext,
+  TResolvedTypesMeta,
+  // @ts-ignore
+  TEventsCausingServices = TResolvedTypesMeta['eventsCausingServices'],
+  // @ts-ignore
+  TIndexedEvents = TResolvedTypesMeta['indexedEvents']
+> = {
+  [K in keyof TEventsCausingServices]?: InvokeCreator<
+    TContext,
+    // @ts-ignore
+    TIndexedEvents[TEventsCausingServices[K]]
+  >;
+};
+
+type MakeKeysRequired<T extends string> = { [K in T]: unknown };
+
+type MaybeMakeMissingImplementationsRequired<
+  TImplementationType,
+  TMissingImplementations,
+  TRequireMissingImplementations
+> = TRequireMissingImplementations extends true
+  ? IsNever<TMissingImplementations> extends true
+    ? {}
+    : {
+        [K in Cast<TImplementationType, string>]: MakeKeysRequired<// @ts-ignore
+        TMissingImplementations>;
+      }
+  : {};
+
+type GenerateActionsConfigPart<
+  TContext,
+  TResolvedTypesMeta,
+  TRequireMissingImplementations,
+  // @ts-ignore
+  TMissingActions = TResolvedTypesMeta['missingImplementations']['actions']
+> = MaybeMakeMissingImplementationsRequired<
+  'actions',
+  TMissingActions,
+  TRequireMissingImplementations
+> & {
+  actions?: TypegenMachineOptionsActions<TContext, TResolvedTypesMeta>;
+};
```

Comment:
This looks a little bit complicated - the idea is that we make all named implementations optional and only "mark" required ones as required using `unknown` value.

This might look like a trick but it makes perfect sense if you think about what a type intersection is - a resulting type has to satisfy **all** member types. So by making the property required, even with `unknown` used as a value, we make it required because an intersection of "it has to be here and it might be here" is "it has to be here". Check a simple demo here:
[TS playground](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAggYxASwPYDsDOMC8MDewMxMAZqqgPwBcMAFFAG634zjS0DkAYgPK+cYAXwCUuAHwxGqZABNgQ4KEiwASlACOAV2QAnKLMQoM2PIRJkKtLegDW6VAHd0Cpe1jdk6AIYAbI2hYuPBIgdgAZDDq2noGASZKslAIvt76ZDahGGTodN5ZWLSePv4FmCK00nJKpLn4osC1dKzkqLR0YjiS9cIiQA)

This is somewhat a workaround for not being able to conditionally change the required/optional modifier in a single mapped type.

This also has a somewhat unfortunate effect that `undefined` stays assignable to the property:
[TS playground](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAggYxASwPYDsDOMC8MDewMxMAZqqgPwBcMAFFAG634zjS0DkAYgPK+cYAXwCUuAHwxGqZABNgQ4KEiwASlACOAV2QAnKLMQoM2PIRJkKtLegDW6VAHd0Cpe1jdk6AIYAbI2hYuPBIgdgAZDDq2noGASZKslAIvt76ZDahGGTodN5ZWLSePv4FmCK00nJKpLn4osC1dKzkqLR0YjiS9cIijXWWbTA2SaReBr1AA)

However, I feel like this is not something that can easily trip people - we don't support passing `undefined` at those positions anywhere and it's not an easy mistake to make. I plan to add a runtime dev-only validation for this which I think is a good enough solution. In addition to that, we could validate this in the VScode extension so I don't think it's a big deal. By not handling this at the type level we avoid creating two mapped types with conditionals etc that would intersect with each other here - so I think that it's worth it.

The issue also actually goes away with `exactOptionalPropertyTypes: true`:
[TS playground](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true#code/FAFwngDgpgBAggYxASwPYDsDOMC8MDewMxMAZqqgPwBcMAFFAG634zjS0DkAYgPK+cYAXwCUuAHwxGqZABNgQ4KEiwASlACOAV2QAnKLMQoM2PIRJkKtLegDW6VAHd0Cpe1jdk6AIYAbI2hYuPBIgdgAZDDq2noGASZKslAIvt76ZDahGGTodN5ZWLSePv4FmCK00nJKpLn4osC1dKzkqLR0YjiS9cIijXWWbTA2SaReBr1AA)

But we can't expect everyone to turn this on - and this option has not been included in the `strict` settings by the TS team.